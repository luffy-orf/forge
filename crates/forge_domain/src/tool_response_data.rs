use std::collections::HashMap;
use serde::{Deserialize, Serialize};

/// Represents the structured data for tool responses.
#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(tag = "type")]
pub enum ToolResponseData {
    /// File read operation response
    #[serde(rename = "file_read")]
    FileRead {
        /// Path of the file that was read
        path: String,
        /// Total number of lines in the file
        total_lines: Option<usize>,
        /// Additional metadata specific to file read
        #[serde(flatten)]
        metadata: HashMap<String, serde_json::Value>,
    },
    
    /// File write operation response
    #[serde(rename = "file_write")]
    FileWrite {
        /// Path of the file that was written
        path: String,
        /// Total bytes written
        bytes_written: Option<usize>,
        /// Whether the operation was an update to an existing file
        was_update: Option<bool>,
        /// Additional metadata specific to file write
        #[serde(flatten)]
        metadata: HashMap<String, serde_json::Value>,
    },
    
    /// Shell command execution response
    #[serde(rename = "shell")]
    Shell {
        /// The command that was executed
        command: String,
        /// Exit code of the command
        exit_code: Option<i32>,
        /// Additional metadata specific to shell commands
        #[serde(flatten)]
        metadata: HashMap<String, serde_json::Value>,
    },
    
    /// Patch operation response (for file modifications)
    #[serde(rename = "patch")]
    Patch {
        /// Path of the file that was patched
        path: String,
        /// Total characters in the patched file
        total_chars: Option<usize>,
        /// Warning message if any syntax issues were detected
        warning: Option<String>,
        /// Additional metadata specific to patch operations
        #[serde(flatten)]
        metadata: HashMap<String, serde_json::Value>,
    },
    
    /// Generic response for tools without specific structured data
    #[serde(rename = "generic")]
    Generic {
        /// Tool-specific metadata
        #[serde(flatten)]
        metadata: HashMap<String, serde_json::Value>,
    },
}

impl ToolResponseData {
    /// Create a FileRead response
    pub fn file_read(path: impl Into<String>) -> Self {
        Self::FileRead {
            path: path.into(),
            total_lines: None,
            metadata: HashMap::new(),
        }
    }
    
    /// Create a FileWrite response
    pub fn file_write(path: impl Into<String>) -> Self {
        Self::FileWrite {
            path: path.into(),
            bytes_written: None,
            was_update: None,
            metadata: HashMap::new(),
        }
    }
    
    /// Create a Shell response
    pub fn shell(command: impl Into<String>) -> Self {
        Self::Shell {
            command: command.into(),
            exit_code: None,
            metadata: HashMap::new(),
        }
    }
    
    /// Create a Patch response
    pub fn patch(path: impl Into<String>) -> Self {
        Self::Patch {
            path: path.into(),
            total_chars: None,
            warning: None,
            metadata: HashMap::new(),
        }
    }
    
    /// Create a Generic response
    pub fn generic() -> Self {
        Self::Generic {
            metadata: HashMap::new(),
        }
    }
    
    /// Add a value to the metadata
    pub fn with_metadata(mut self, key: impl Into<String>, value: impl Into<serde_json::Value>) -> Self {
        match &mut self {
            Self::FileRead { metadata, .. } => {
                metadata.insert(key.into(), value.into());
            }
            Self::FileWrite { metadata, .. } => {
                metadata.insert(key.into(), value.into());
            }
            Self::Shell { metadata, .. } => {
                metadata.insert(key.into(), value.into());
            }
            Self::Patch { metadata, .. } => {
                metadata.insert(key.into(), value.into());
            }
            Self::Generic { metadata } => {
                metadata.insert(key.into(), value.into());
            }
        }
        self
    }
    
    /// Update the total lines for FileRead
    pub fn with_total_lines(mut self, total_lines: usize) -> Self {
        if let Self::FileRead { total_lines: t, .. } = &mut self {
            *t = Some(total_lines);
        }
        self
    }
    
    /// Update the bytes written for FileWrite
    pub fn with_bytes_written(mut self, bytes_written: usize) -> Self {
        if let Self::FileWrite { bytes_written: b, .. } = &mut self {
            *b = Some(bytes_written);
        }
        self
    }
    
    /// Update the was_update flag for FileWrite
    pub fn with_was_update(mut self, was_update: bool) -> Self {
        if let Self::FileWrite { was_update: w, .. } = &mut self {
            *w = Some(was_update);
        }
        self
    }
    
    /// Update the exit code for Shell
    pub fn with_exit_code(mut self, exit_code: i32) -> Self {
        if let Self::Shell { exit_code: e, .. } = &mut self {
            *e = Some(exit_code);
        }
        self
    }
    
    /// Update the total chars for Patch
    pub fn with_total_chars(mut self, total_chars: usize) -> Self {
        if let Self::Patch { total_chars: t, .. } = &mut self {
            *t = Some(total_chars);
        }
        self
    }
    
    /// Update the warning for Patch
    pub fn with_warning(mut self, warning: impl Into<String>) -> Self {
        if let Self::Patch { warning: w, .. } = &mut self {
            *w = Some(warning.into());
        }
        self
    }
    
    /// Convert the ToolResponseData to YAML front matter format
    /// 
    /// The front matter contains all the metadata, while the content is provided separately.
    /// This follows the format:
    /// ```text
    /// ---
    /// key1: value1
    /// key2: value2
    /// ---
    /// Content goes here
    /// ```
    pub fn to_front_matter(&self, content: impl Into<String>) -> String {
        let content_str = content.into();
        
        // Convert the enum to a serializable representation
        let data = serde_json::to_value(self).unwrap_or_default();
        
        // Build the front matter manually
        let mut result = String::from("---\n");
        
        if let serde_json::Value::Object(map) = data {
            // Use serde_yml to convert the JSON object to YAML
            let yaml = serde_yml::to_string(&map).unwrap_or_default();
            result.push_str(&yaml);
        }
        
        result.push_str("---\n");
        result.push_str(&content_str);
        
        result
    }
    
    /// Parse front matter format back into a ToolResponseData and content
    pub fn from_front_matter(text: &str) -> (Option<Self>, String) {
        // Simple parsing using string operations rather than relying on gray_matter
        if !text.starts_with("---\n") {
            return (None, text.to_string());
        }
        
        // Find the end of the front matter
        if let Some(end_index) = text[4..].find("\n---\n") {
            let yaml_end = 4 + end_index;
            let yaml_content = &text[4..yaml_end];
            let content = &text[(yaml_end + 5)..]; // Skip past the ending delimiter
            
            // Parse the YAML
            if let Ok(value) = serde_yml::from_str::<serde_json::Value>(yaml_content) {
                if let Ok(data) = serde_json::from_value::<Self>(value) {
                    return (Some(data), content.to_string());
                }
            }
            
            // Return just the content if we couldn't parse the YAML
            return (None, content.to_string());
        }
        
        // If we couldn't find the ending delimiter, return the original text
        (None, text.to_string())
    }
} 