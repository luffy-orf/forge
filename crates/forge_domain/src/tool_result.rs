use derive_setters::Setters;
use serde::{Deserialize, Serialize};

use crate::{ToolCallFull, ToolCallId, ToolName, tool_response_data::ToolResponseData};

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, Setters)]
#[setters(strip_option, into)]
pub struct ToolResult {
    pub name: ToolName,
    pub call_id: Option<ToolCallId>,
    #[setters(skip)]
    pub content: String,
    #[setters(skip)]
    pub is_error: bool,
    #[setters(skip)]
    #[serde(skip)]
    pub data: Option<ToolResponseData>,
}

impl ToolResult {
    pub fn new(name: ToolName) -> ToolResult {
        Self {
            name,
            call_id: None,
            content: String::default(),
            is_error: false,
            data: None,
        }
    }

    pub fn success(mut self, content: impl Into<String>) -> Self {
        self.content = content.into();
        self.is_error = false;
        self
    }

    pub fn failure(mut self, err: anyhow::Error) -> Self {
        let mut output = String::new();
        output.push_str("\nERROR:\n");

        for cause in err.chain() {
            output.push_str(&format!("Caused by: {cause}\n"));
        }

        self.content = output;
        self.is_error = true;
        self
    }
    
    pub fn with_data(mut self, data: ToolResponseData) -> Self {
        self.data = Some(data);
        self
    }
    
    /// Helper method to set both the content and structured data in one call
    /// This also returns the content directly, making it convenient for tools to generate their response
    pub fn with_frontmatter_response(mut self, data: ToolResponseData, content: impl Into<String>) -> Self {
        let content_str = content.into();
        self.content = content_str;
        self.data = Some(data);
        self
    }
}

impl From<ToolCallFull> for ToolResult {
    fn from(value: ToolCallFull) -> Self {
        Self {
            name: value.name,
            call_id: value.call_id,
            content: String::default(),
            is_error: false,
            data: None,
        }
    }
}

impl std::fmt::Display for ToolResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // If we have ToolResponseData, use front matter format
        if let Some(data) = &self.data {
            write!(f, "{}", data.to_front_matter(&self.content))
        } else {
            // Legacy XML format for backward compatibility
            write!(f, "<forge_tool_result>")?;
            write!(
                f,
                "<forge_tool_name>{}</forge_tool_name>",
                self.name.as_str()
            )?;
            let content = format!("<![CDATA[{}]]>", self.content);
            if self.is_error {
                write!(f, "<e>{content}</e>")?;
            } else {
                write!(f, "<success>{content}</success>")?;
            }

            write!(f, "</forge_tool_result>")
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use serde_json::json;

    use super::*;
    use crate::tool_response_data::ToolResponseData;

    #[test]
    fn test_snapshot_minimal() {
        let result = ToolResult::new(ToolName::new("test_tool"));
        assert_snapshot!(result);
    }

    #[test]
    fn test_snapshot_full() {
        let result = ToolResult::new(ToolName::new("complex_tool"))
            .call_id(ToolCallId::new("123"))
            .failure(anyhow::anyhow!(
                json!({"key": "value", "number": 42}).to_string()
            ));
        assert_snapshot!(result);
    }

    #[test]
    fn test_snapshot_with_special_chars() {
        let result = ToolResult::new(ToolName::new("xml_tool")).success(
            json!({
                "text": "Special chars: < > & ' \"",
                "nested": {
                    "html": "<div>Test</div>"
                }
            })
            .to_string(),
        );
        assert_snapshot!(result);
    }

    #[test]
    fn test_display_minimal() {
        let result = ToolResult::new(ToolName::new("test_tool"));
        assert_snapshot!(result.to_string());
    }

    #[test]
    fn test_display_full() {
        let result = ToolResult::new(ToolName::new("complex_tool"))
            .call_id(ToolCallId::new("123"))
            .success(
                json!({
                    "user": "John Doe",
                    "age": 42,
                    "address": [{"city": "New York"}, {"city": "Los Angeles"}]
                })
                .to_string(),
            );
        assert_snapshot!(result.to_string());
    }

    #[test]
    fn test_display_special_chars() {
        let result = ToolResult::new(ToolName::new("xml_tool")).success(
            json!({
                "text": "Special chars: < > & ' \"",
                "nested": {
                    "html": "<div>Test</div>"
                }
            })
            .to_string(),
        );
        assert_snapshot!(result.to_string());
    }

    #[test]
    fn test_success_and_failure_content() {
        let success = ToolResult::new(ToolName::new("test_tool")).success("success message");
        assert!(!success.is_error);
        assert_eq!(success.content, "success message");

        let failure =
            ToolResult::new(ToolName::new("test_tool")).failure(anyhow::anyhow!("error message"));
        assert!(failure.is_error);
        assert_eq!(failure.content, "\nERROR:\nCaused by: error message\n");
    }
    
    #[test]
    fn test_frontmatter_format() {
        let data = ToolResponseData::file_read("/path/to/file.txt")
            .with_total_lines(100)
            .with_metadata("encoding", "utf-8");
            
        let result = ToolResult::new(ToolName::new("file_read"))
            .success("File content here")
            .with_data(data);
            
        let output = result.to_string();
        assert!(output.starts_with("---"));
        assert!(output.contains("type: file_read"));
        assert!(output.contains("path: /path/to/file.txt"));
        assert!(output.contains("total_lines: 100"));
        assert!(output.contains("encoding: utf-8"));
        assert!(output.contains("File content here"));
    }
    
    #[test]
    fn test_parse_frontmatter() {
        let input = r#"---
type: file_read
path: /path/to/file.txt
total_lines: 100
encoding: utf-8
---
File content here"#;
        
        let (data, content) = ToolResponseData::from_front_matter(input);
        assert!(data.is_some());
        if let Some(ToolResponseData::FileRead { path, total_lines, metadata }) = data {
            assert_eq!(path, "/path/to/file.txt");
            assert_eq!(total_lines, Some(100));
            assert_eq!(metadata.get("encoding").unwrap().as_str().unwrap(), "utf-8");
        } else {
            panic!("Expected FileRead data");
        }
        assert_eq!(content, "File content here");
    }

    #[test]
    fn test_complete_frontmatter_example() {
        // Create a structured data object for a file read operation
        let data = ToolResponseData::file_read("/example/file.txt")
            .with_total_lines(42)
            .with_metadata("encoding", "utf-8")
            .with_metadata("file_size", 1024);
            
        // Create a tool result with the structured data
        let result = ToolResult::new(ToolName::new("forge_tool_fs_read"))
            .call_id(ToolCallId::new("abc123"))
            .with_frontmatter_response(data, "This is the content of the file\nSecond line\nThird line");
            
        // Convert to string to get the front matter format
        let output = result.to_string();
        
        // Verify the output contains expected fields
        assert!(output.contains("---"));
        assert!(output.contains("type: file_read"));
        assert!(output.contains("path: /example/file.txt"));
        assert!(output.contains("total_lines: 42"));
        assert!(output.contains("encoding: utf-8"));
        assert!(output.contains("file_size: 1024"));
        assert!(output.contains("---"));
        assert!(output.contains("This is the content of the file"));
        assert!(output.contains("Second line"));
        assert!(output.contains("Third line"));
        
        // Parse back from string
        let (parsed_data, content) = ToolResponseData::from_front_matter(&output);
        assert!(parsed_data.is_some());
        if let Some(ToolResponseData::FileRead { path, total_lines, metadata }) = parsed_data {
            assert_eq!(path, "/example/file.txt");
            assert_eq!(total_lines, Some(42));
            assert_eq!(metadata.get("encoding").unwrap().as_str().unwrap(), "utf-8");
            assert_eq!(metadata.get("file_size").unwrap().as_u64().unwrap(), 1024);
        } else {
            panic!("Expected FileRead data");
        }
        
        assert_eq!(content, "This is the content of the file\nSecond line\nThird line");
    }
}
