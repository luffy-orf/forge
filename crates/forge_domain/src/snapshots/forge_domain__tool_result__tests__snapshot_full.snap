---
source: crates/forge_domain/src/tool_result.rs
expression: result
---
<forge_tool_result><forge_tool_name>complex_tool</forge_tool_name><e><![CDATA[
ERROR:
Caused by: {"key":"value","number":42}
]]></e></forge_tool_result>
